### PR TITLE
Added Serde support

### PR DESCRIPTION
**Pull request by TylerBloom**
This adds serialization and deserialization support via serde. This is done through an intermediate type that wraps each instance of `T` with its count.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,8 @@ maintenance = { status = "passively-maintained" }
 
 [dependencies]
 griddle = { version = "0.5", optional = true }
+serde = { version = "1.0", features = ["derive"], optional = true }
 
 [features]
 amortize = ["griddle"]
+with_serde = ["serde"]
```
Comment (jonhoo) [inline Cargo.toml]: Why not just have the feature be called `serde`? I think that will just work without adding anything to `[features]`.

File: src/lib.rs [hunk 2]
```diff
@@ -24,7 +24,8 @@ use std::collections::hash_map::RandomState;
 #[cfg(not(feature = "amortize"))]
 use std::collections::HashMap;
 use std::hash::{BuildHasher, Hash};
-
+#[cfg(feature = "with_serde")]
+mod with_serde;
```
Comment (jonhoo) [inline src/lib.rs]:
```suggestion
mod with_serde;

```

File: src/with_serde.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+impl<'de, T> Visitor<'de> for WrapperVisitor<T>
+where
+    T: Deserialize<'de>,
+{
+    type Value = EntryWrapper<T>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("struct Duration")
```
Comment (jonhoo) [inline src/with_serde.rs]: This doesn't look right?

File: src/with_serde.rs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+#[derive(Deserialize)]
+#[serde(field_identifier, rename_all = "lowercase")]
+enum Field {
+    Entry,
+    Count,
+}
+
+//struct WrapperVisitor;
```
Comment (jonhoo) [inline src/with_serde.rs]:
```suggestion
```

File: src/with_serde.rs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+impl<'de, T, S> Deserialize<'de> for HashBag<T, S>
+where
+    T: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(HashBagVisitor::<T, S>::new())
```
Comment (jonhoo) [inline src/with_serde.rs]: Hmm, wouldn't this make more sense using `deserialize_map`, and having `T` be the key and `count` be the value? That'd also save you the wrapper type.

**Review by jonhoo — CHANGES_REQUESTED** [review]
Thanks for taking the time to do this! I left a few notes inline.

File: Cargo.toml (hunk 1 above)
Comment (TylerBloom) [reply Cargo.toml]: It would probably be easy to have a feature called `serde` but when I tried `serde = ["serde"]`, cargo wasn't particularly pleased. However, I know that it (or something close enough) *can* be done. I just wasn't sure how.

File: src/with_serde.rs (hunk 3 above)
Comment (TylerBloom) [reply src/with_serde.rs]: Sorry, copypasta. Such be `"struct HashBag EntryWrapper"` or something along those lines.

File: src/with_serde.rs (hunk 5 above)
Comment (TylerBloom) [reply src/with_serde.rs]: Keys to maps need to be string-like (at least in JSON), although serde will not complain about this at compile time. I had the same idea and ran into issues when doing some testing with JSON. That is to say, the following struct would produce the following invalid JSON.

```rust
struct MyStruct {
  foo: String,
}
```

```json
{
  {"foo": "bar"} : 42,
  {"foo": "baz"} : 84,
}
```

File: Cargo.toml (hunk 1 above)
Comment (jonhoo) [reply Cargo.toml]: Just don't list it under `[features]` at all, and that should do the trick :) Optional dependencies implicitly create a feature by the same name.

File: src/with_serde.rs (hunk 5 above)
Comment (jonhoo) [reply src/with_serde.rs]: Oh, interesting. But at the same time, surely `HashMap` has the same problem if you try to use a key that isn't legal in key position for JSON?

Comment (TylerBloom) [reply src/with_serde.rs]: Ya, after some testing, it would seem that a regular hashmap with keys that aren't strings (or at least that aren't json primitives) can't be jsonified. It also doesn't seem to help if the type implements `ToString` and/or `From<String>`/`From<&str>`.

I would imagine (with no way of verifying) that serde is intended to map Rust's map as closely to JSON's maps as possible. However, the fact that `HashBag` uses a map seems to be of little consequence to the user. In fact, `HashBag` could be implemented using a `HashSet<Entry<T>>` where `Entry` encapsulates an instance of `T` and its count and uses its held instance of `T` for `Hash` and `Eq`.

File: Cargo.toml [hunk 6]
```diff
@@ -12,13 +12,16 @@ repository = "https://github.com/jonhoo/hashbag.git"
 keywords = ["bag","multiset","set"]
 categories = ["data-structures"]
 
+serde = ["serde"]
```
Comment (jonhoo) [inline Cargo.toml]: This should be removed

File: src/lib.rs [hunk 7]
```diff
@@ -24,7 +24,8 @@ use std::collections::hash_map::RandomState;
 #[cfg(not(feature = "amortize"))]
 use std::collections::HashMap;
 use std::hash::{BuildHasher, Hash};
-
```
Comment (jonhoo) [inline src/lib.rs:27]: Let's leave an empty line here between the `use` and `mod`s.

File: src/serde.rs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    H: BuildHasher + Clone,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut bag = serializer.serialize_seq(Some(self.len()))?;
+
+        for (entry, count) in self.set_iter() {
+            bag.serialize_element(&EntryWrapper { entry, count })?;
```
Comment (jonhoo) [inline src/serde.rs]: I think this ends up serializing `&T`, not `T`. Now, that's not really a problem since, I believe, `Serialize` is implemented for `&T` too, but wasn't immediately obvious to me.

File: src/serde.rs [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        bag.end()
+    }
+}
+
+pub(crate) struct EntryWrapper<T> {
```
Comment (jonhoo) [inline src/serde.rs]: Even if you don't like going the `deserialize_map` route, which I'm okay with, I don't think we actually need this. We could instead just serialize using `(T, usize)`, and that way we won't need any of the bits below. _And_ we'd end up with a more compact JSON representation.

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 10]
```diff
@@ -25,6 +25,8 @@ use std::collections::hash_map::RandomState;
 use std::collections::HashMap;
 use std::hash::{BuildHasher, Hash};
 
+#[cfg(feature = "serde")]
+mod serde;
```
Comment (jonhoo) [inline src/lib.rs:33]:
```suggestion
mod serde;

```

File: src/serde.rs [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    fn visit_seq<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: SeqAccess<'de>,
+    {
+        let mut bag: HashBag<T, S> =
+            HashBag::with_capacity_and_hasher(access.size_hint().unwrap_or(0), Default::default());
+
+        while let Some(entry) = access.next_element::<(T, usize)>()? {
+            bag.insert_many(entry.0, entry.1);
```
Comment (jonhoo) [inline src/serde.rs:45]: I wonder if this should error if the key already exists (by inspecting the return value of `insert_many` — what do you think?

File: src/serde.rs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    H: BuildHasher + Clone,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut bag = serializer.serialize_seq(Some(self.len()))?;
+
+        for (entry, count) in self.set_iter() {
+            bag.serialize_element(&(&entry, &count))?;
```
Comment (jonhoo) [inline src/serde.rs]:
```suggestion
            bag.serialize_element(&(entry, &count))?;
```
`entry` is already a `&T` — no need to take a further reference.

File: src/serde.rs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    use serde_json;
+
+    #[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+    struct VeryHelpfulStruct {
+        pub(crate) name: String,
+    }
+
+    #[test]
+    fn format_simple_data() {
+        let vikings: HashBag<String> = ["Einar", "Olaf", "Harald"]
```
Comment (jonhoo) [inline src/serde.rs]: We should include at least one item with a count > 1 to make sure we do the right thing in that case too.

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/serde.rs (hunk 11 above)
Comment (TylerBloom) [reply src/serde.rs:45]: That depends. What assumptions do we want to make about the incoming data? My assumptions would be that the keys were properly `T` (de)serialized (i.e. their implementations are correct) and the user is aware of the possibility of duplicates. I think we should follow the behavior of other hashed collections. Neither `HashMap` nor `HashSet` returns an error if there are repeat keys/values. Lastly, I don't think we should provide an outcome that isn't in the `HashBag` API. There is not a, say, `swap` method that swaps the count of a key with another count.

Putting this all together, I think we should assume (with documentation) that the user ought to be aware of the potential for duplicate keys in their data, and we should treat these duplicates like the other hashed collections do, by using existing insert methods. Anything more specialized than that is up to the user (for example, take only the last occurrence of a key/count pair).

Comment (jonhoo) [reply src/serde.rs:45]: That seems reasonable to me!

File: src/serde.rs (hunk 13 above)
Comment (TylerBloom) [reply src/serde.rs]: Good call.

File: src/serde.rs [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        let reconsituted_vikings: HashBag<VeryHelpfulStruct> =
+            serde_json::from_str(&jsonified_vikings).expect("Unable to convert json to hashbag!");
+        println!("{:?}", reconsituted_vikings);
+        assert_eq!(vikings, reconsituted_vikings);
+    }
+
+    #[test]
+    fn repeat_entries() {
+        let vikings: HashBag<VeryHelpfulStruct> =
+            ["Einar", "Olaf", "Olaf", "Harald", "Harald", "Harald"]
```
Comment (jonhoo) [inline src/serde.rs:120]: I don't think we need this to be a separate test — lets instead just have the two existing tests also have duplicates.

File: src/serde.rs [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    H: BuildHasher + Clone,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut bag = serializer.serialize_seq(Some(self.len()))?;
+
+        for (entry, count) in self.set_iter() {
+            bag.serialize_element(&(entry, &count))?;
```
Comment (jonhoo) [inline src/serde.rs]: The `count` is a `usize`, which is `Copy`, so it doesn't need the `&` either :)

File: src/lib.rs [hunk 16]
```diff
@@ -12,6 +12,10 @@
 //! If you want to use a hash table with [amortized resizes](https://github.com/jonhoo/griddle/),
 //! set the `amortize` feature.
 //!
+//! (De)serialization via serde is also available with the `serde` feature.
+//! Serialization note: if the incoming data has repeated keys, the resulting `HashBag` will merge
```
Comment (jonhoo) [inline src/lib.rs]: Would you mind adding a test for this behavior too?

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 16 above)
Comment (TylerBloom) [reply src/lib.rs]: I added two tests to mirror the other tests. I also corrected my wording here. We are not concerned about communicating how we serialize a `HashBag`, only how we deserialize one.

File: src/serde.rs [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+        let reconsituted_vikings: HashBag<VeryHelpfulStruct> =
+            serde_json::from_str(&jsonified_vikings).expect("Unable to convert json to hashbag!");
+        println!("From json: {:?}", reconsituted_vikings);
+        assert_eq!(vikings, reconsituted_vikings);
+    }
+
+    #[test]
+    fn repeat_simple_entries() {
+        let jsonified_vikings: String =
+            "[[\"Einar\",1],[\"Olaf\",1],[\"Olaf\",1],[\"Harald\",2],[\"Harald\",1]]".to_string();
```
Comment (jonhoo) [inline src/serde.rs:151]: Just a tip: you can use [raw string literals](https://doc.rust-lang.org/reference/tokens.html#raw-string-literals) to avoid all the `\"` in there :)

**Review by jonhoo — APPROVED** [review]
Excellent, thanks!

**Comment by jonhoo** [comment]
Published in 0.1.5 :tada:
